app/src/org/commcare/views/widgets/GregorianDateWidget.java

```diff
@@ -0,0 +1,386 @@
+            yearText.setText(String.valueOf(calendar.get(Calendar.YEAR)));
+        }
+    }
+
+    //Checks if all text fields contain valid values, corrects fields with invalid values, updates calendar based on text fields.
+    protected void validateTextOnButtonPress(){
+        validateDayTextOnButtonPress();
+
+        monthArrayPointer = monthSpinner.getSelectedItemPosition();
+        calendar.set(Calendar.MONTH, monthArrayPointer);
+
+        String yearTextValue = yearText.getText().toString();
+        calendar.set(Calendar.YEAR, Integer.parseInt(yearTextValue));
+    }
+
+    //Day-specific validation. Refactored into a separate method because it's called during month selection.
+    private void validateDayTextOnButtonPress() {
+        String dayTextString = dayText.getText().toString();
+        int dayTextValue = Integer.parseInt(dayTextString);
+        int maxDayOfMonth = calendar.getActualMaximum(Calendar.DAY_OF_MONTH);
+
+        if(dayTextValue >= maxDayOfMonth){
+            dayTextValue = calendar.getActualMaximum(Calendar.DAY_OF_MONTH);
+            dayText.setText(String.valueOf(dayTextValue));
+        }else if(dayTextValue < 1){
+            dayTextValue = calendar.getActualMinimum(Calendar.DAY_OF_MONTH);
+            dayText.setText(String.valueOf(dayTextValue));
+        }
+
+        calendar.set(Calendar.DAY_OF_MONTH, dayTextValue);
+    }
+
+    @Override
+    protected UniversalDate decrementMonth(long millisFromJavaEpoch) {
+        DateTime dt = new DateTime(millisFromJavaEpoch).minusMonths(1);
+        return constructUniversalDate(dt);
+    }
+
+    @Override
+    protected UniversalDate decrementYear(long millisFromJavaEpoch) {
+        DateTime dt = new DateTime(millisFromJavaEpoch).minusYears(1);
+        return constructUniversalDate(dt);
+    }
+
+    @Override
+    protected UniversalDate fromMillis(long millisFromJavaEpoch) {
+        return constructUniversalDate(new DateTime(millisFromJavaEpoch));
+    }
+
+    @Override
+    protected UniversalDate incrementMonth(long millisFromJavaEpoch) {
+        DateTime dt = new DateTime(millisFromJavaEpoch).plusMonths(1);
+        return constructUniversalDate(dt);
+    }
+
+    @Override
+    protected UniversalDate incrementYear(long millisFromJavaEpoch) {
+        DateTime dt = new DateTime(millisFromJavaEpoch).plusYears(1);
+        return constructUniversalDate(dt);
+    }
+
+    @Override
+    protected String[] getMonthsArray() {
+        calendar = Calendar.getInstance();
+
+        String [] monthNames = new String[12];
+        final Map<String, Integer> monthMap = calendar.getDisplayNames(Calendar.MONTH, Calendar.LONG, Locale.getDefault());
+        monthList = new ArrayList<>(monthMap.keySet());
+            Collections.sort(monthList, new Comparator<String>(){
+                @Override
+                public int compare(String a, String b){
+                    return monthMap.get(a) - monthMap.get(b);
+                }
+            });

```

Comment:
I think you have similar sort code in `CalendarFragment`. Could probably be lifted to a static helper.